### ISSUE TITLE
undefined symbol: runtime.fastrand while building Gio

### ISSUE THREAD

**Issue by eliasnaur** [tinygo-org/tinygo#2663]
I'm investigating whether [Gio](https://gioui.org) can run on TinyGo. After replacing references to [UnsafeAddr](https://pkg.go.dev/reflect#Value.UnsafeAddr), I ran into

```
$ tinygo version
tinygo version 0.22.0-dev darwin/amd64 (using go version go1.17.7 and LLVM version 11.1.0)
$ tinygo build -o dapdap.hex -target stm32f469disco gio.go
ld.lld: error: undefined symbol: runtime.fastrand
>>> referenced by maphash.go:210 (/usr/local/go/src/hash/maphash/maphash.go:210)
>>>               /var/folders/3p/_lbqc_4n5t764s1c44sql13w0000gn/T/tinygo2067000345/main.o:(runtime.run$1$gowrapper)
>>> referenced by maphash.go:211 (/usr/local/go/src/hash/maphash/maphash.go:211)
>>>               /var/folders/3p/_lbqc_4n5t764s1c44sql13w0000gn/T/tinygo2067000345/main.o:(runtime.run$1$gowrapper)
error: failed to link /var/folders/3p/_lbqc_4n5t764s1c44sql13w0000gn/T/tinygo2067000345/main: exit status 1
```

Gio.go:

```go
package main

import (
	"gioui.org/io/router"
	"gioui.org/op"
	"gioui.org/op/paint"
)

func main() {
	var ops op.Ops
	var r router.Router
	paint.PaintOp{}.Add(&ops)
	r.Frame(&ops)
}
```

I faintly recall an issue around maps with `interface{}` keys (`map[interface{}]...`) which Gio makes extensive use of. Perhaps the missing `runtime.fastrand` symbol is related to that?

**Comment by aykevl** [comment]
> I faintly recall an issue around maps with `interface{}` keys (`map[interface{}]...`) which Gio makes extensive use of.

This was fixed a long time ago. But yes, this used to be an issue.

> Perhaps the missing `runtime.fastrand` symbol is related to that?

Probably not (either way). It simply needs to be implemented in the runtime (the signature is `func fastrand() uint32`). If you want, you can make a PR for it. We already have code you could reuse for this purpose:

https://github.com/tinygo-org/tinygo/blob/a680bfbb7afcb424c4d8110255336f85fd66e7e6/testdata/gc.go#L3-L16

Or maybe I can take a look at fixing it myself, it shouldn't be very difficult.

**Comment by aykevl** [comment]
Take a look here: https://github.com/tinygo-org/tinygo/pull/2664

**Comment by eliasnaur** [comment]
Thanks a lot! With #2664, I can confirm Gio builds successfully (haven't tried to run it yet).

**Comment by eliasnaur** [comment]
Fixed by #2664 .
